### PR TITLE
DPP-1616 bulk download agreements

### PR DESCRIPTION
**Pull request by scottywakefield**

### CODE REVIEW

File: src/SFA.DAS.ProviderApprenticeshipsService.Web/Views/Agreement/Agreements.cshtml [hunk 1]
```diff
@@ -39,6 +39,8 @@
                 }
             </tbody>
         </table>
+        <p>You can download this data as a <a href="@Url.Action("DownloadCsv", "Agreement")" aria-label="">CSV</a>
```
Comment (chrisfoster76) [inline src/SFA.DAS.ProviderApprenticeshipsService.Web/Views/Agreement/Agreements.cshtml]: We should probably slip in a couple of aria-labels giving the full link purpose, eg. "Download agreement data in CSV format".

File: src/SFA.DAS.ProviderApprenticeshipsService.Web/Helpers/CsvWriterFacade.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Collections;
+using System.IO;
+using System.Web.Mvc;
+using CsvHelper;
+
+namespace SFA.DAS.ProviderApprenticeshipsService.Web.Helpers
+{
+    public static class CsvWriterFacade
```
Comment (chrisfoster76) [inline src/SFA.DAS.ProviderApprenticeshipsService.Web/Helpers/CsvWriterFacade.cs]: Does this facade have to be static. Could we not instantiate the facade, and compose it of the CsvWriter? For the Excel implementation, we instantiate the workbook and Deliver() it.

File: src/SFA.DAS.ProviderApprenticeshipsService.Web/Helpers/CsvWriterFacade.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        private const string CsvContentType = "text/csv";
+
+        public static FileStreamResult Deliver(IEnumerable records, string fileName)
+        {
+            var memoryStream = new MemoryStream();
```
Comment (chrisfoster76) [inline src/SFA.DAS.ProviderApprenticeshipsService.Web/Helpers/CsvWriterFacade.cs]: We ought to _use_ these streams and writers, etc. and get disposal for free.

File: src/SFA.DAS.ProviderApprenticeshipsService.Web/Views/Agreement/Agreements.cshtml (hunk 1 above)
Comment (scottywakefield) [reply src/SFA.DAS.ProviderApprenticeshipsService.Web/Views/Agreement/Agreements.cshtml]: good idea.

File: src/SFA.DAS.ProviderApprenticeshipsService.Web.UnitTests/Orchestrators/Formatters/GivenACsvFormatter.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+using System.Text;
+using FluentAssertions;
+using NUnit.Framework;
+using SFA.DAS.ProviderApprenticeshipsService.Web.Models.Agreement;
+using SFA.DAS.ProviderApprenticeshipsService.Web.Orchestrators.Formatters;
+
+namespace SFA.DAS.ProviderApprenticeshipsService.Web.UnitTests.Orchestrators.Formatters
+{
+    [TestFixture]
+    public class GivenACsvFormatter
```
Comment (chrisfoster76) [inline src/SFA.DAS.ProviderApprenticeshipsService.Web.UnitTests/Orchestrators/Formatters/GivenACsvFormatter.cs]: I had thought we'd cover the ExcelFormatter in a unit test too? With ExcelFormatter as the sut, we'd pass a list of agreements to it just the same as here, but then we'd take the byte array output and run it back through the XLWorkbook constructor as a stream, asserting that the worksheet has got all the stuff we'd expect?

**Pull request merged**